geth/node/txqueue_manager.go

```diff
@@ -0,0 +1,360 @@
+		m.NotifyOnQueuedTxReturn(queuedTx, ErrInvalidCompleteTxSender)
+		return gethcommon.Hash{}, ErrInvalidCompleteTxSender
+	}
+
+	config, err := m.nodeManager.NodeConfig()
+	if err != nil {
+		return gethcommon.Hash{}, err
+	}
+
+	// Send the transaction finally.
+	var hash gethcommon.Hash
+	var txErr error
+
+	if config.UpstreamConfig.Enabled {
+		hash, txErr = m.completeRemoteTransaction(queuedTx, password)
+	} else {
+		hash, txErr = m.completeLocalTransaction(queuedTx, password)
+	}
+
+	// when incorrect sender tries to complete the account,
+	// notify and keep tx in queue (so that correct sender can complete)
+	if txErr == keystore.ErrDecrypt {
+		m.NotifyOnQueuedTxReturn(queuedTx, txErr)
+		return hash, txErr
+	}
+
+	log.Info("finally complete transaction", "id", queuedTx.ID)
+
+	queuedTx.Hash = hash
+	queuedTx.Err = txErr
+	queuedTx.Done <- struct{}{} // sendTransaction() waits on this, notify so that it can return
+
+	return hash, txErr
+}
+
+func (m *TxQueueManager) completeLocalTransaction(queuedTx *common.QueuedTx, password string) (gethcommon.Hash, error) {
+	log.Info("complete transaction using local node", "id", queuedTx.ID)
+
+	les, err := m.nodeManager.LightEthereumService()
+	if err != nil {
+		return gethcommon.Hash{}, err
+	}
+
+	return les.StatusBackend.SendTransaction(context.Background(), status.SendTxArgs(queuedTx.Args), password)
+}
+
+func (m *TxQueueManager) completeRemoteTransaction(queuedTx *common.QueuedTx, password string) (gethcommon.Hash, error) {
+	log.Info("complete transaction using upstream node", "id", queuedTx.ID)
+
+	var emptyHash gethcommon.Hash
+
+	config, err := m.nodeManager.NodeConfig()
+	if err != nil {
+		return emptyHash, err
+	}
+
+	selectedAcct, err := m.accountManager.SelectedAccount()
+	if err != nil {
+		return emptyHash, err
+	}
+
+	client, err := m.nodeManager.RPCClient()
+	if err != nil {
+		return emptyHash, err
+	}
+
+	// We need to request a new transaction nounce from upstream node.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	var txCount hexutil.Uint
+	if err := client.CallContext(ctx, &txCount, "eth_getTransactionCount", queuedTx.Args.From, "pending"); err != nil {
+		return emptyHash, err
+	}
+
+	chainID := big.NewInt(int64(config.NetworkID))
+	nonce := uint64(txCount)
+	gas := (*big.Int)(queuedTx.Args.Gas)
+	gasPrice := (*big.Int)(queuedTx.Args.GasPrice)
+	dataVal := []byte(queuedTx.Args.Data)
+	priceVal := (*big.Int)(queuedTx.Args.Value)
+
+	tx := types.NewTransaction(nonce, *queuedTx.Args.To, priceVal, gas, gasPrice, dataVal)
+	signedTx, err := types.SignTx(tx, types.NewEIP155Signer(chainID), selectedAcct.AccountKey.PrivateKey)
+	if err != nil {
+		return emptyHash, err
+	}
+
+	txBytes, err := rlp.EncodeToBytes(signedTx)
+	if err != nil {
+		return emptyHash, err
+	}
+
+	ctx2, cancel2 := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel2()
+
+	var result json.RawMessage
```

Comment:
What happens to result here, seems like it's used to receive response but not saved, or returned anywhere?

Reply:
Actually, we don't need to compare it as it's not possible it will change. Following https://github.com/ethereum/go-ethereum/blob/bc0e6a5e68904b2a1251240b85dd97dc8fa07e11/ethclient/ethclient.go#L426-L432